### ISSUE TITLE
support for the bidirectional GRU

### ISSUE THREAD

**Issue by xhzhao** [SeanNaren/deepspeech.pytorch#215]
We know that the DS2 paper use bidirectional GRU/LSTM, and we found that the paper add the bidirectional result after each layer. But i could not found this add operation in this pytorch implementation. Should we follow the paper or not?
Right now, for the second GRU layer, the input size=1600, hidden size=800.
I'm wondering if this is the root cause of the accuracy issue.
![image](https://user-images.githubusercontent.com/17486215/34825857-21e79a3a-f70f-11e7-9162-c23202bbcc0a.png)

**Comment by ryanleary** [comment]
This happens here: https://github.com/SeanNaren/deepspeech.pytorch/blob/master/model.py#L70

**Comment by xhzhao** [comment]
@ryanleary thanks for your response. you are right.

**Comment by xhzhao** [comment]
@ryanleary but i have another question about the pretrained model size.
from my calculation, the default model is about 152MB, but the downloaded the pretrained mode is about 291MB. Is there any wrong about my calculation?
![image](https://user-images.githubusercontent.com/17486215/34865232-d93819dc-f7b2-11e7-8105-8bdc0939cf06.png)

**Comment by ryanleary** [comment]
Youre not taking into consideration the weights for the reverse direction of the RNNs. The parameters weights are not shared in this implementation

**Comment by SeanNaren** [comment]
Going to close this for now...

**Comment by xhzhao** [comment]
@ryanleary thanks for your remainder, but i really take the direction into account (D in my formula). take the last GRU layer as an example, parameter count = D*(I+H)*3H = 2*(800+800)*3*800=7680000, weight is float type, so the weight is 7680000*4/1e6 MB= 30.72 MB.

**Comment by ryanleary** [comment]
@xhzhao yes you are correct -- sorry I missed that term. The numbers you calculated line up with the in-memory requirements. 

We're apparently shipping in the final models the optimizer state_dict as well as the model parameters which will allow continued training from where the training left off. Removing this component from the package using something like this:

```python
package = torch.load("/path/to/librispeech_pretrained.pth", map_location=lambda storage, loc: storage)
del package['optim_dict']
torch.save(package, "/path/to/librispeech_pretrained_no_optim.pth")
```

will result in the model shrinking considerably to be around the size you quoted:

```bash
ls -lah | grep librispeech_pretrained
-rw-r--r--    1 ryanleary  staff   291M Jan 15 23:18 librispeech_pretrained.pth
-rw-r--r--    1 ryanleary  staff   145M Jan 15 23:21 librispeech_pretrained_no_optim.pth
```

cc @SeanNaren

**Comment by xhzhao** [comment]
@ryanleary thanks for your help, 145M really match with my formula, 3805776*4/1024/1024= 145.18MB
